### ISSUE TITLE
NPE exception when reading from queue when old queue data files are deleted

### ISSUE THREAD

**Issue by riteshadval** [OpenHFT/Chronicle-Queue#534]
Hi I am trying to implement periodic old queue file purge logic using StoreFileListener. I am using latest release net.openhft:chronicle-queue:5.16.13. The issue I am having is: after  the roll cycle file gets rolled because of next cycle is active, I delete the queue file which just got released in StoreFileListener, then I create a new tailer and try to read message. It throws below NPE:

The same happens if I create a brand new queue pointing to same queue dir and create a tailer.

seeing below NPE exception when trying to implement queue file purging logic:

java.lang.NullPointerException
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.inACycle(SingleChronicleQueueExcerpts.java:1198)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:1000)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:942)
	at net.openhft.chronicle.wire.MarshallableIn.readText(MarshallableIn.java:95)
	at com.test.edge.api.queue.TestDeleteQueueFile.testQueueFileDeletionWhileInUse(TestDeleteQueueFile.java:133)
	
I am attaching a standalone unit test which exhibit this issue
:
[TestDeleteQueueFile.txt](https://github.com/OpenHFT/Chronicle-Queue/files/2390526/TestDeleteQueueFile.txt)

also inline below:

Not sure if something silly on my part but not able to identify whats the issue.

`
    package com.test.edge.api.queue;
    
    import net.openhft.chronicle.core.time.SetTimeProvider;
    import net.openhft.chronicle.queue.ChronicleQueue;
    import net.openhft.chronicle.queue.ExcerptAppender;
    import net.openhft.chronicle.queue.ExcerptTailer;
    import net.openhft.chronicle.queue.RollCycles;
    import net.openhft.chronicle.queue.impl.StoreFileListener;
    import net.openhft.chronicle.queue.impl.single.SingleChronicleQueueBuilder;
    import org.apache.commons.io.FileUtils;
    import org.junit.After;
    import org.junit.Assert;
    import org.junit.Before;
    import org.junit.Test;
    
    import java.io.File;
    import java.nio.file.Files;
    import java.nio.file.Path;
    import java.nio.file.Paths;
    import java.util.ArrayList;
    import java.util.HashMap;
    import java.util.List;
    import java.util.Map;
    
    public class TestDeleteQueueFile {
    
        private Path tempQueueDir;
    
        @Before
        public void setUp() throws Exception {
            tempQueueDir = Files.createTempDirectory("unitTestQueueDir");
        }
    
        @After
        public void tearDown() throws Exception {
            FileUtils.deleteDirectory(tempQueueDir.toFile());
            Assert.assertFalse(tempQueueDir.toFile().exists());
            System.out.println("Deleted " + tempQueueDir.toFile().getAbsolutePath());
        }
    
        @Test
        public void testQueueFileDeletionWhileInUse() throws Exception {
            SetTimeProvider timeProvider = new SetTimeProvider();
    
            String queueName = "unitTestQueue";
    
            QueueStoreFileListener listener = new QueueStoreFileListener(queueName);
    
            try (ChronicleQueue queue = SingleChronicleQueueBuilder.binary(tempQueueDir + "/" + queueName).
                    timeProvider(timeProvider).storeFileListener(listener)
                    .build()) {
    
                ExcerptAppender appender = queue.acquireAppender();
    
                System.out.println("first index : " + queue.firstIndex());
                Assert.assertEquals(Long.MAX_VALUE, queue.firstIndex());
    
                //write 10 records should go to first day file
                for (int i = 0; i < 10; i++) {
                    appender.writeText("test");
                }
    
                long indexAfter10Records = appender.lastIndexAppended();
                System.out.println("index after writing 10 records: " + indexAfter10Records);
    
                //roll to next day file
                timeProvider.advanceMillis(24 * 60 * 60 * 1000);
    
    
                //write 5 records in next file
                for (int i = 0; i < 5; i++) {
                    appender.writeText("test2");
                }
    
                Map<String, List> queueToRollFilesOnAcquireMap = listener.getQueueToRollFilesOnAcquireMap();
                Map<String, List> queueToRollFilesOnReleaseMap = listener.getQueueToRollFilesOnReleaseMap();
    
                Assert.assertEquals(1, queueToRollFilesOnAcquireMap.size());
                List<String> files = queueToRollFilesOnAcquireMap.get(queueName);
                Assert.assertEquals(1, files.size());
                String secondFile = files.get(0);
    
                //other will have 1 as only first file is released
                files = queueToRollFilesOnReleaseMap.get(queueName);
                Assert.assertEquals(1, files.size());
                String firstFile = files.get(0);
    
                Assert.assertNotEquals(firstFile, secondFile);
    
    
                long firstIndex = queue.firstIndex();
    
    
                long indexAfter5Records = appender.lastIndexAppended();
                System.out.println("index after writing 5 records: " + indexAfter5Records);
    
                //now lets create one reader which will read all content
                ExcerptTailer excerptTailer = queue.createTailer();
                for (int i = 0; i < 10; i++) {
                    Assert.assertEquals("test", excerptTailer.readText());
                }
    
                System.out.println("index after reading 10 records: " + excerptTailer.index());
                Assert.assertEquals(firstIndex, excerptTailer.index() - 10);
                for (int i = 0; i < 5; i++) {
                    Assert.assertEquals("test2", excerptTailer.readText());
                }
    
                System.out.println("index after reading 5 records: " + excerptTailer.index());
                Assert.assertEquals(indexAfter5Records, excerptTailer.index() - 1);
    
                //lets delete first file
                System.out.println("Deleting first release file: " + firstFile);
                Files.delete(Paths.get(firstFile));
    
    
                long firstIndex3 = queue.firstIndex();
                Assert.assertEquals(firstIndex, firstIndex3);
    
                // and create a tailer it should only read
                //data in second file
                ExcerptTailer excerptTailer2 = queue.createTailer();
                System.out.println("index before reading 5: " + excerptTailer2.index());
    
                //AFTER CREATING A BRAND NEW TAILER, BELOW ASSERTION ALSO FAILS
                //WAS EXPECTING THAT TAILER CAN READ FROM START OF QUEUE BUT INDEX IS LONG.MAX
    
                //Assert.assertEquals(indexAfter5Records -5, excerptTailer2.index() -1);
    
                //BELOW THROWS NPE, WAS EXPECTING THAT WE CAN READ FROM SECOND DAILY QUEUE FILE
                System.out.println("excerptTailer2: " + excerptTailer2.peekDocument());
                for (int i = 0; i < 5; i++) {
                    Assert.assertEquals("test2", excerptTailer2.readText());
                }
    
                //SAME ERROR WHEN CREATING A BRAND NEW QUEUE AND TRYING TO READ IT
    
    //            // create brand new queue and read see how it behaves
    //            ChronicleQueue queue2 =  SingleChronicleQueueBuilder.binary(tempQueueDir + "/" + queueName).
    //                    timeProvider(timeProvider).storeFileListener(listener)
    //                    .build();
    //
    //            long firstIndex2 = queue2.firstIndex();
    //            Assert.assertNotEquals(firstIndex, firstIndex2);
    //            Assert.assertNotEquals(indexAfter5Records, firstIndex2);
    //
    //            ExcerptTailer excerptTailer3 = queue.createTailer();
    //            for (int i = 0; i < 5; i++) {
    //                Assert.assertEquals("test2", excerptTailer3.readText());
    //            }
    
            }
    
        }
    
        final class QueueStoreFileListener implements StoreFileListener {
    
            private String queueName;
            private Map<String, List> queueToRollFilesOnReleaseMap = new HashMap<>();
            private Map<String, List> queueToRollFilesOnAcquireMap = new HashMap<>();
    
            public QueueStoreFileListener(String queueName) {
                this.queueName = queueName;
            }
    
            @Override
            public void onReleased(int cycle, File file) {
                System.out.println("onReleased called cycle: " + cycle + "file: " + file);
    
                List<String> files = queueToRollFilesOnReleaseMap.get(queueName);
                if (files == null) {
                    files = new ArrayList<>();
                }
    
                String fileAbsPath = file.getAbsolutePath();
                if (!files.contains(fileAbsPath)) {
                    files.add(fileAbsPath);
                }
                queueToRollFilesOnReleaseMap.put(queueName, files);
    
                //update acquire file map
                List<String> acqfiles = queueToRollFilesOnAcquireMap.get(queueName);
                acqfiles.remove(file.getAbsolutePath());
                queueToRollFilesOnAcquireMap.put(queueName, acqfiles);
    
            }
    
            @Override
            public void onAcquired(int cycle, File file) {
                System.out.println("onAcquired called cycle: " + cycle + "file: " + file);
    
                List<String> files = queueToRollFilesOnAcquireMap.get(queueName);
                if (files == null) {
                    files = new ArrayList<>();
                }
    
                String fileAbsPath = file.getAbsolutePath();
                if (!files.contains(fileAbsPath)) {
                    files.add(fileAbsPath);
                }
    
                queueToRollFilesOnAcquireMap.put(queueName, files);
    
    
            }
    
            public Map<String, List> getQueueToRollFilesOnAcquireMap() {
                return queueToRollFilesOnAcquireMap;
            }
    
            public Map<String, List> getQueueToRollFilesOnReleaseMap() {
                return queueToRollFilesOnReleaseMap;
            }
        }
    }
`

**Comment by riteshadval** [comment]
Anyone looking at this ticket?

**Comment by dpisklov** [comment]
@riteshadval We will look into that as soon as we are able to. None of our paid customers experienced this issue (and also it's not quite the scenario majority of our users ever come across), and we prioritise the issues from our paid customers over the issues raised by the community. Pull requests are welcome though.
If you are interested in SLA for getting issues fixed please consider getting a support contract with us, for that please get in contact with us at sales@chronicle.software

**Comment by RobAustin** [comment]
@riteshadval after you delete the file, can you create another instance of the ChronicleQueue and use that instead, I suspect that it has cached the existence of the file that you deleted, so creating a new instance of ChronicleQueue and use that, should work around this problem. I'd be grateful if you could let me know if this works. As this will indicate to us what the problem is. Also if this works perhaps we could work together to get this resolved. I may be able to point you at the area of code to change in chronicle-queue so that you could create the fix and issue a pull request.

**Comment by RobAustin** [comment]
we have added the following method:

```
/**
* call this method if you delete a file from a chronicle-queue directory
*/
net.openhft.chronicle.queue.ChronicleQueue#refreshDirectlyListing
```

The problem that we have is that we cache the structure of your directory this is because hitting the file system adds latency. If you call this method, if you delete files, then it will update our caches accordingly,

I've added your test and updated it slightly

```
package net.openhft.chronicle.queue;

import net.openhft.chronicle.core.time.SetTimeProvider;
import net.openhft.chronicle.queue.impl.StoreFileListener;
import net.openhft.chronicle.queue.impl.single.SingleChronicleQueueBuilder;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TestDeleteQueueFile {

    private Path tempQueueDir;

    @Before
    public void setUp() throws Exception {
        tempQueueDir = Files.createTempDirectory("unitTestQueueDir");
    }

    @After
    public void tearDown() {

        DirectoryUtils.deleteDir(tempQueueDir.toFile());
        Assert.assertFalse(tempQueueDir.toFile().exists());
        System.out.println("Deleted " + tempQueueDir.toFile().getAbsolutePath());
    }

    @Test
    public void testQueueFileDeletionWhileInUse() throws Exception {
        SetTimeProvider timeProvider = new SetTimeProvider();

        String queueName = "unitTestQueue";

        QueueStoreFileListener listener = new QueueStoreFileListener(queueName);

        try (ChronicleQueue queue = SingleChronicleQueueBuilder.binary(tempQueueDir + "/" + queueName).
                timeProvider(timeProvider).storeFileListener(listener)
                .build()) {

            ExcerptAppender appender = queue.acquireAppender();

            System.out.println("first index : " + queue.firstIndex());
            Assert.assertEquals(Long.MAX_VALUE, queue.firstIndex());

            //write 10 records should go to first day file
            for (int i = 0; i < 10; i++) {
                appender.writeText("test");
            }

            long firstIndex = queue.firstIndex();

            long indexAfter10Records = appender.lastIndexAppended();
            System.out.println("index after writing 10 records: " + indexAfter10Records);

            //roll to next day file
            timeProvider.advanceMillis(24 * 60 * 60 * 1000);

            appender.writeText("test2");
            long firstIndexOfSecondCycle = appender.lastIndexAppended();

            //write 5 records in next file
            for (int i = 0; i < 4; i++) {
                appender.writeText("test2");
            }

            Map<String, List> queueToRollFilesOnAcquireMap = listener.getQueueToRollFilesOnAcquireMap();
            Map<String, List> queueToRollFilesOnReleaseMap = listener.getQueueToRollFilesOnReleaseMap();

            Assert.assertEquals(1, queueToRollFilesOnAcquireMap.size());
            List<String> files = queueToRollFilesOnAcquireMap.get(queueName);
            Assert.assertEquals(1, files.size());
            String secondFile = files.get(0);

            //other will have 1 as only first file is released
            files = queueToRollFilesOnReleaseMap.get(queueName);
            Assert.assertEquals(1, files.size());
            String firstFile = files.get(0);

            Assert.assertNotEquals(firstFile, secondFile);

            long indexAfter5Records = appender.lastIndexAppended();
            System.out.println("index after writing 5 records: " + indexAfter5Records);

            //now lets create one reader which will read all content
            ExcerptTailer excerptTailer = queue.createTailer();
            for (int i = 0; i < 10; i++) {
                Assert.assertEquals("test", excerptTailer.readText());
            }

            System.out.println("index after reading 10 records: " + excerptTailer.index());
            Assert.assertEquals(firstIndex, excerptTailer.index() - 10);
            for (int i = 0; i < 5; i++) {
                Assert.assertEquals("test2", excerptTailer.readText());
            }

            System.out.println("index after reading 5 records: " + excerptTailer.index());
            Assert.assertEquals(indexAfter5Records, excerptTailer.index() - 1);

            //lets delete first file
            System.out.println("Deleting first release file: " + firstFile);

            Files.delete(Paths.get(firstFile));

            queue.refreshDirectlyListing();

            Assert.assertEquals(queue.firstIndex(), firstIndexOfSecondCycle);

            // and create a tailer it should only read
            //data in second file
            ExcerptTailer excerptTailer2 = queue.createTailer();
            System.out.println("index before reading 5: " + excerptTailer2.index());

            //AFTER CREATING A BRAND NEW TAILER, BELOW ASSERTION ALSO FAILS
            //WAS EXPECTING THAT TAILER CAN READ FROM START OF QUEUE BUT INDEX IS LONG.MAX

            Assert.assertEquals(indexAfter5Records -5, excerptTailer2.index() -1);

            //BELOW THROWS NPE, WAS EXPECTING THAT WE CAN READ FROM SECOND DAILY QUEUE FILE
            System.out.println("excerptTailer2: " + excerptTailer2.peekDocument());
            for (int i = 0; i < 5; i++) {
                Assert.assertEquals("test2", excerptTailer2.readText());
            }


        }

    }

    final class QueueStoreFileListener implements StoreFileListener {

        private String queueName;
        private Map<String, List> queueToRollFilesOnReleaseMap = new HashMap<>();
        private Map<String, List> queueToRollFilesOnAcquireMap = new HashMap<>();

        public QueueStoreFileListener(String queueName) {
            this.queueName = queueName;
        }

        @Override
        public void onReleased(int cycle, File file) {
            System.out.println("onReleased called cycle: " + cycle + "file: " + file);

            List<String> files = queueToRollFilesOnReleaseMap.get(queueName);
            if (files == null) {
                files = new ArrayList<>();
            }

            String fileAbsPath = file.getAbsolutePath();
            if (!files.contains(fileAbsPath)) {
                files.add(fileAbsPath);
            }
            queueToRollFilesOnReleaseMap.put(queueName, files);

            //update acquire file map
            List<String> acqfiles = queueToRollFilesOnAcquireMap.get(queueName);
            acqfiles.remove(file.getAbsolutePath());
            queueToRollFilesOnAcquireMap.put(queueName, acqfiles);

        }

        @Override
        public void onAcquired(int cycle, File file) {
            System.out.println("onAcquired called cycle: " + cycle + "file: " + file);

            List<String> files = queueToRollFilesOnAcquireMap.get(queueName);
            if (files == null) {
                files = new ArrayList<>();
            }

            String fileAbsPath = file.getAbsolutePath();
            if (!files.contains(fileAbsPath)) {
                files.add(fileAbsPath);
            }

            queueToRollFilesOnAcquireMap.put(queueName, files);

        }

        public Map<String, List> getQueueToRollFilesOnAcquireMap() {
            return queueToRollFilesOnAcquireMap;
        }

        public Map<String, List> getQueueToRollFilesOnReleaseMap() {
            return queueToRollFilesOnReleaseMap;
        }
    }
}
```
Thank you for reporting this, the test case was very helpfull.

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.17.1 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), BOM-2.17.44 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.44)

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.17.1 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), BOM-2.17.46 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.46)

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.17.1 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), BOM-2.17.48 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.48)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.50](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.50)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.52](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.52)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.53](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.53)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.60](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.60)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.72](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.72)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.1](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.1), [BOM-2.17.74](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.74)
